### PR TITLE
Edge-hanging corners in mesh

### PR DESCRIPTION
**Pull request by hannesbrandt**
# p4est_mesh: add corner neighbors across coarse edges

Following up on issue https://github.com/cburstedde/p4est/issues/244 and the test proposed by @scottaiton in https://github.com/cburstedde/p4est/pull/235.
We introduce a `p4est_mesh_params_t` that contains all mesh-creation parameters with the possibility to add further parameters in the future. The `p4est_mesh_t` now contains a `p4est_mesh_params_t` instance to store the parameters it was created with.
In 3D, we add a new `edgehanging_corners` flag to the set of parameters. When set to true, quadrants that are corner neighbors across a coarse edge are stored in the same format as all other corner-neighbor relations of the mesh. If set to false, the mesh is created as usual.
Furthermore, we add `p4est_mesh_new_params` which creates a mesh according to a passed `p4est_mesh_params_t` instance. A default parameter instance can be created with `p4est_mesh_params_new` and destroyed with `p4est_mesh_params_destroy`.
Apart from this, we remove a few unreachable lines of code in `mesh_iter_edge` and add some documentation in `p4est_mesh.h`. Especially, `p4est_mesh_get_neighbors` is now marked as outdated.

### CODE REVIEW

File: src/p4est_mesh.c [hunk 1]
```diff
@@ -162,6 +171,16 @@ mesh_edge_process_inter_tree_edges (p8est_iter_edge_info_t * info,
   equads = P4EST_ALLOC (p4est_locidx_t, nAdjacentQuads);
   eedges = P4EST_ALLOC (int8_t, nAdjacentQuads);
 
+  ccorners = NULL;
+  cquads = NULL;
+  if (mesh->params.edgehanging_corners == 1 &&
```
Comment (cburstedde) [inline src/p4est_mesh.c:176]: Please check booleans with if (a) or if (!a). The C standard considers any nonzero as true.

File: src/p4est_mesh.c [hunk 2]
```diff
@@ -554,56 +602,22 @@ mesh_iter_edge (p8est_iter_edge_info_t * info, void *user_data)
     return;
   }
   if (cz == 2) {
+    P4EST_ASSERT (info->tree_boundary);
```
Comment (cburstedde) [inline src/p4est_mesh.c:605]: Good catch!

File: src/p4est_mesh.c [hunk 3]
```diff
@@ -1136,6 +1212,36 @@ p4est_mesh_memory_used (p4est_mesh_t * mesh)
   return all_memory;
 }
 
+static void
+mesh_params_init (p4est_mesh_params_t * params)
```
Comment (cburstedde) [inline src/p4est_mesh.c:1216]: I'd be suggesting to work with static parameters. Then we wouldn't need the new and destroy functions, just p4est_mesh_params_init, which first memsets the whole structure to zero and then writes defaults into it.

File: src/p4est_mesh.c [hunk 4]
```diff
@@ -1147,6 +1253,23 @@ p4est_mesh_t       *
 p4est_mesh_new_ext (p4est_t * p4est, p4est_ghost_t * ghost,
                     int compute_tree_index, int compute_level_lists,
                     p4est_connect_type_t btype)
+{
+  p4est_mesh_params_t parameters;
+
```
Comment (cburstedde) [inline src/p4est_mesh.c:1258]: If we call p4est_mesh_params_init (&mesh->parameters) we avoid duplicate settings.

Comment (cburstedde) [reply src/p4est_mesh.c:1241]: You're right; I had a slightly different angle. The `mesh->params = *params;` is fine.

I was thinking of using after `/* initialize parameter struct to pass to mesh_new_params */`
something like `p4est_mesh_init_params (&parameters)` and thereafter only overwriting the values that are not default. Would that be safer or do you prefer the current approach?

As a side note, shall we rename any leftover `parameters` to `params` for consistency?

Comment (hannesbrandt) [reply src/p4est_mesh.c:1241]: That's a good point, initializing first with `p4est_mesh_init_params` ensures that new additions to the `mesh_params_t` do not lead to unexpected behaviour in `mesh_new_ext`. I've changed this and renamed `parameters` to `params` as well.

Comment (cburstedde) [reply src/p4est_mesh.c:1241]: Looks very nice!

**Pull request merged**